### PR TITLE
feat(server-auth-testing): add more flexible request matching to the OPA rule

### PR DESCRIPTION
**Pull request by dhenneke**
This change adds support for checking whether the JWT was correctly sent to the OPA. In the old implementation one could use `onRequest(method, path)` to configure an expected request. Now there is a builder approach `onRequest().withMethod(method).withPath(path).withJwt(jwt)` that also allows to check for explicit JWT values. This is also possible for verifying requests: `verify(1, onRequest().withMethod(method).withPath(path).withJwt(jwt))`. There was no functional change in the existing methods.

The existing `onRequest(method, path)` approach was also marked as deprecated since the new approach gives much more flexibility. Futhermore, it should be advised to always check if the JWT is if correctly forwarded to the OPA.

### CODE REVIEW

File: sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/opa/testing/OpaRule.java [hunk 1]
```diff
@@ -75,18 +86,59 @@ protected void after() {
...
+  public interface RequestMethodBuilder {
+    RequestPathBuilder withMethod(String method);
+  }
+
+  public interface RequestPathBuilder {
+    RequestExtraBuilder withPath(String path);
+  }
+
+  public interface RequestExtraBuilder extends AllowBuilder {
+    RequestExtraBuilder withJwt(String jwt);
```
Comment (JoergSiebahn) [inline sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/opa/testing/OpaRule.java]: Why does it return the `RequestExtraBuilder`? This seems to allow multiple calls to `withJwt(…)` in the builder. Does that makes sense?

File: sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/opa/testing/OpaRule.java [hunk 2]
```diff
@@ -75,18 +86,59 @@ protected void after() {
...
   public void mock(BuildBuilder builder) {
     builder.build(wire);
   }
 
-  private RequestPatternBuilder matchForVerify(String httpMethod, String[] paths) {
-    return RequestPatternBuilder.newRequestPattern(RequestMethod.POST, urlMatching("/.*"))
-        .withRequestBody(matchingJsonPath("$.input.httpMethod", equalTo(httpMethod)))
-        .withRequestBody(matchingJsonPath("$.input.path", equalToJson(getJson(paths))));
+  public interface RequestMethodBuilder {
+    RequestPathBuilder withMethod(String method);
```
Comment (JoergSiebahn) [inline sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/opa/testing/OpaRule.java]: Maybe `httpMethod` is a more specific term?

**Comment by JoergSiebahn** [comment]
As far as I can see, `path` in the tests contains always a single path segment. Do you have any tests that verify the implementation of the `splitPath`-Method? There may be another edge case with trailing slash. Would it be fine to split and filter for not blank? Maybe you can use the algorithm of the `OpaAuthFilter`. Following the code, it uses `decodePath` of Jerseys `UriComponent` which is static.

File: sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/opa/testing/OpaRule.java (hunk 1 above)
Comment (dhenneke) [reply sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/opa/testing/OpaRule.java]: No, that would not make sense. In future, we could also add the possibility to check for e.g. request headers. We could then extend this interface.

File: sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/opa/testing/OpaRule.java (hunk 2 above)
Comment (dhenneke) [reply sda-commons-server-auth-testing/src/main/java/org/sdase/commons/server/opa/testing/OpaRule.java]: I'll change it

**Comment by dhenneke** [comment]
> As far as I can see, `path` in the tests contains always a single path segment. Do you have any tests that verify the implementation of the `splitPath`-Method? There may be another edge case with trailing slash. Would it be fine to split and filter for not blank?

Fair point. I'll check the tests for the examples.
